Created a feature_extraction folder and added feature description.md (#91)
johko/computer-vision-course: chapters/en/Unit 1 - Fundamentals/feature_extraction/feature_description.md

﻿# **Feature Description**

## **How can we represent features in data structures?**

> Representing features in data structures is crucial for organizing and manipulating data effectively. Features, also known as attributes or variables, can be diverse, ranging from numerical values and categories to more complex structures like images or text. Here are some ways to represent features in data structures:

Reviewer: I think this section should focus more on visual features only. I also didn't understand what the term "data structures" refers to. 

Generally speaking, I would mention that features try to find some patterns in the visual (edges, corners, points of interest, etc.) which are helpful in solving high-level tasks later on. I liked the "descriptor" term that you used in the next section. That would be great to put it here in the introduction as well.

> 

Reviewer 2: I feel like some of the below data types are never meant for images anyway, it's best to have only image processing specific data types not to confuse the user. we don't have to elaborate bottom up if there's redundant information

Reviewer 2: Yes, some of the data types that aren't used in computer vision can be removed

1. **One hot encoding:** Transforming categorical variables into binary vectors where each bit represents a category.
1. **Text features**
1. **Bag of Words Representation:** Breaking text into individual words and counting their occurrences.
1. **Word Embeddings:** Representing words as dense vectors capturing semantic meanings.
1. **Image features**
1. **Pixel Values:** Strong pixel values matrices or multi-dimensional arrays.
1. **Convolutional Neural Network features(CNN):** Extracting features using pre-trained CNN models.

## **What Makes a good descriptor**

A good descriptor in the context of image processing or computer vision is a set of characteristics or features that effectively represent key information about an object or scene in an image. Here are some aspects that contribute to making a good descriptor:

1. **Invariant to transformation:** Descriptors should ideally be robust to variations like rotation, translation, scaling and changes in illumination. This means that regardless of how an object is positioned or how the image is altered, the descriptor should remain relatively unchanged and contain the same description as the original one.

> 1. **Distinctivness:** A good descriptor captures unique information about the object. It should be able to discriminate between different objects or parts of image and be distinct enough to differentiate them from similar elements.

Reviewer 2: ```suggestion
2. **Distinctiveness:** A good descriptor captures unique information about the object. It should be able to discriminate between different objects or parts of the image and be distinct enough to differentiate them from similar elements.
```

> 1. **Dimensionality:**  Good descriptors often have a manageable size, conveying enough information without being excessively large. Balancing dimensionality is crucial for efficiency in processing and storage. 

Reviewer 2: ```suggestion
3. **Dimensionality:**  Good descriptors often have a manageable size, conveying enough information without being excessively large. Balancing dimensionality is crucial for efficiency in processing and storage. 
```

6. **Compatibility with Matching Algorithms:**  Descriptors are often used with matching algorithms to find correspondences between different images. A good descriptor should be suitable for the matching algorithm, whether based on distance metrics, machine learning models, or other techniques.
7. **Computational Efficiency:** Efficiency is crucial to a descriptor, especially in real-time applications. Descriptors should be computationally feasible for quick processing, particularly in scenarios where speed is crucial, such as robotics or autonomous vehicles.
8. **Adaptability:** Descriptors that can adapt or learn from the data they are processing can be highly effective, especially in situations where the characteristics of the objects or scenes may change over time. This can increase the useability of the descriptor.
9. **Noise Robustness:** Descriptors should be able to handle noise in the image data without significantly compromising their ability to accurately represent the underlying features.

## Some Techniques used in Feature Descriptors

> 1. **SIFT :**  It stands for scale invariant feature transform. It is widely used algorithm in computer vision and image processing for detecting and describing local features in images.

Reviewer: I think it would be good to consider using h3 headers for SIFT and SURF sections below instead of nesting a list in both cases. It would be visually more appealing. Also "scale invariant feature transform" should start with capital letters, as it's an acronym expansion.

﻿# **Feature Description**

> ## **How can we represent features in Images?**

Author: Ok I will keep an eye on this from now on

> ## **How can we represent features in Images?**

Reviewer 2: @aman06012003 it would be nice if this was fixed and we can resolve the comment

Representing features in images is crucial for organizing and manipulating image effectively. Features, also known as attributes or variables, can be diverse, ranging from numerical values and categories to more complex structures like images or text. Here are some ways to represent features in image:

> 1. **Numerical features**

Author: sorry I didn't got your point :)

> 1. **Numerical features**

Reviewer 3: I think what Maria means is, that right now it is a continuing list, but actually is should be multiple small lists like
- Numerical features
  - Arrays/Lists
  - Tensors
- Categorical Features
  - Dictionaries/Lists
  - One Hot Encoding

> 1. **Numerical features**

Author: Ok

6. **Compatibility with Matching Algorithms:**  Descriptors are often used with matching algorithms to find correspondences between different images. A good descriptor should be suitable for the matching algorithm, whether based on distance metrics, machine learning models, or other techniques.
7. **Computational Efficiency:** Efficiency is crucial to a descriptor, especially in real-time applications. Descriptors should be computationally feasible for quick processing, particularly in scenarios where speed is crucial, such as robotics or autonomous vehicles.
8. **Adaptability:** Descriptors that can adapt or learn from the data they are processing can be highly effective, especially in situations where the characteristics of the objects or scenes may change over time. This can increase the useability of the descriptor.
9. **Noise Robustness:** Descriptors should be able to handle noise in the image data without significantly compromising their ability to accurately represent the underlying features.

## Some Techniques used in Feature Descriptors

>  ### SIFT :

Reviewer 2: ```suggestion
### SIFT
```

  It stands for scale invariant feature transform. It is widely used algorithm in computer vision and image processing for detecting and describing local features in images.

The working of SIFT is given below:
1. **Scale Space Extrema detection:** It starts by detecting potential interest points in an image across multiple scales. It looks for locations in the image where the difference of Gaussian function reaches a maximum or minimum over space and scale. These keypoint locations are considered stable under various scale changes.

> 2. **Keypoint Localization:** Once potential keypoints are identified, SIFT refines their positions to sub-pixel accuracy and discards low-contrast key points and keypoints on edges to ensure accurate localization.** 

Reviewer 3: ```suggestion
2. **Keypoint Localization:** Once potential keypoints are identified, SIFT refines their positions to sub-pixel accuracy and discards low-contrast key points and keypoints on edges to ensure accurate localization.
```
